tokio/src/sync/batch_semaphore.rs

```diff
@@ -189,44 +188,58 @@ impl Semaphore {
+        let mut lock = Some(waiters);
+        let mut is_empty = false;
+        while rem > 0 {
+            let mut waiters = lock.take().unwrap_or_else(|| self.waiters.lock().unwrap());
+            'inner: for slot in &mut wakers[..] {
+                // Was the waiter assigned enough permits to wake it?
+                match waiters.queue.last() {
+                    Some(waiter) => {
+                        if !waiter.assign_permits(&mut rem) {
+                            break 'inner;
+                        }
+                    }
+                    None => {
+                        is_empty = true;
+                        // If we assigned permits to all the waiters in the queue, and there are
+                        // still permits left over, assign them back to the semaphore.
+                        break 'inner;
+                    }
+                };
+                let mut waiter = waiters.queue.pop_back().unwrap();
+                *slot = unsafe { waiter.as_mut().waker.with_mut(|waker| (*waker).take()) };
             }
-            let mut waiter = waiters.queue.pop_back().unwrap();
-            let waker = unsafe { waiter.as_mut().waker.with_mut(|waker| (*waker).take()) };
 
-            waker
-                .expect("if a node is in the wait list, it must have a waker")
-                .wake();
-        }
+            if rem > 0 && is_empty {
+                let permits = rem << Self::PERMIT_SHIFT;
+                assert!(
+                    permits < Self::MAX_PERMITS,
+                    "cannot add more than MAX_PERMITS permits ({})",
+                    Self::MAX_PERMITS
+                );
+                let prev = self.permits.fetch_add(rem << Self::PERMIT_SHIFT, Release);
+                assert!(
+                    prev + permits <= Self::MAX_PERMITS,
+                    "number of added permits ({}) would overflow MAX_PERMITS ({})",
+                    rem,
+                    Self::MAX_PERMITS
+                );
+                rem = 0;
+            }
+
+            drop(waiters); // release the lock
 
-        // If we assigned permits to all the waiters in the queue, and there are
-        // still permits left over, assign them back to the semaphore.
-        if rem > 0 {
-            let permits = rem << Self::PERMIT_SHIFT;
-            assert!(
-                permits < Self::MAX_PERMITS,
-                "cannot add more than MAX_PERMITS permits ({})",
-                Self::MAX_PERMITS
-            );
-            let prev = self.permits.fetch_add(rem << Self::PERMIT_SHIFT, Release);
-            assert!(
-                prev + permits <= Self::MAX_PERMITS,
-                "number of added permits ({}) would overflow MAX_PERMITS ({})",
-                rem,
-                Self::MAX_PERMITS
-            );
+            for slot in &mut wakers[..] {
+                if let Some(waker) = slot.take() {
+                    waker.wake();
+                }
```

Comment:
```rust
else {
    break;
}
```
?

but for the 8waker version it certainly won't matter

Reply:
Initially, this needed to continue looping since a waiter may have had its waker taken but not been removed from the queue yet, meaning that the array could have a `None` followed by more `Some`s. I think that because this method now pops waiters, that won't happen any longer, so yeah, this should probably break.

Reply:
An iterator would probably work here too (not sure if clearer), but not critical either way.

Something like:

```
wakers.iter_mut()
    .filter_map(|w| w.take())
    .for_each(|w| w.wake());
```